deckhand/engine/layering.py

```diff
@@ -0,0 +1,292 @@
+            if last_key in rendered_data and last_key in child_data:
+                rendered_data[last_key] = child_data[last_key]
+            elif last_key in child_data:
+                rendered_data.setdefault(last_key, child_data[last_key])
+            else:
+                # If the key does not exist in the child document, this is a
+                # validation error.
+                raise errors.MissingDocumentKey(
+                    child=child_data, parent=rendered_data, key=last_key)
+
+        return overall_data
+
+    def _find_layering_policy(self):
+        """Retrieve the current layering policy.
+
+        :raises LayeringPolicyMalformed: If the `layerOrder` could not be
+            found in the LayeringPolicy or if it is not a list.
+        :raises LayeringPolicyNotFound: If system has no layering policy.
+        """
+        # TODO(fmontei): There should be a DB call here to fetch the layering
+        # policy from the DB.
+        for doc in self.documents:
+            if doc.all_data['schema'] == self.LAYERING_POLICY_SCHEMA:
+                self.layering_policy = doc
+                break
+
+        if not hasattr(self, 'layering_policy'):
+            raise errors.LayeringPolicyNotFound(
+                schema=self.LAYERING_POLICY_SCHEMA)
+
+        # TODO(fmontei): Rely on schema validation or some such for this.
+        try:
+            self.layer_order = list(self.layering_policy['data']['layerOrder'])
+        except KeyError:
+            raise errors.LayeringPolicyMalformed(
+                schema=self.LAYERING_POLICY_SCHEMA,
+                document=self.layering_policy)
+
+        if not isinstance(self.layer_order, list):
+            raise errors.LayeringPolicyMalformed(
+                schema=self.LAYERING_POLICY_SCHEMA,
+                document=self.layering_policy)
+
+    def _calc_document_children(self):
+        """Determine each document's children.
+
+        For each document, attempts to find the document's children. Adds a new
+        key called "children" to the document's dictionary.
+
+        .. note::
+
+            A document should only have exactly one parent.
+
+            If a document does not have a parent, then its layer must be
+            the topmost layer defined by the `layerOrder`.
+
+        :returns: Ordered list of documents that need to be layered. Each
+            document contains a "parent" in addition to original data. The
+            order highest to lowest layer in `layerOrder`.
+        :raises IndeterminateDocumentParent: If more than one parent document
+            was found for a document.
+        :raises MissingDocumentParent: If the parent document could not be
+            found. Only applies documents with `layeringDefinition` property.
+        """
+        layered_docs = list(
+            filter(lambda x: 'layeringDefinition' in x['metadata'],
+                self.documents))
+
+        all_children = collections.Counter()
```

Comment:
Add comment: "all_children is counter utility for verifying that each document has exactly one parent."